### PR TITLE
NF: QA is a pair instead of a hash map with two elements

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
The "id" was never read, so I can safely remove it

I find far more clear to read a pair than a Map

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java [hunk 1]
```diff
@@ -2250,20 +2250,19 @@ public void updateSearchItemQA() {
...
+            String a = qa.second;
             // Render full question / answer if the bafmt (i.e. "browser appearance") setting forced blank result
-            if ("".equals(qa.get("q")) || "".equals(qa.get("a"))) {
-                HashMap<String, String> qaFull = getCard()._getQA(true, false);
-                if ("".equals(qa.get("q"))) {
-                    qa.put("q", qaFull.get("q"));
+            if ("".equals(q) || "".equals(a)) {
+                Pair<String, String> qaFull = getCard()._getQA(true, false);
+                if ("".equals(q)) {
+                    q = qaFull.first;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:2260]: This appears to have changed functionality, previously the cache (used on subsequent calls) would be updated, now it's not.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java [hunk 2]
```diff
@@ -1112,10 +1112,9 @@ public void updateFieldCache(long[] nids) {
         fields.put("Card", template.getString("name"));
         fields.put(String.format(Locale.US, "c%d", cardNum), "1");
         // render q & a
-        HashMap<String, String> d = new HashMap<>();
-        d.put("id", Long.toString(cid));
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1116]: Why was this removed?

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java:1116]: It is literally never accessed anywhere in the code base. This could actually be seen easily simply because if it were there would be a compilation error since I only get access to first/second

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java (hunk 1 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:2260]: On the one hand, technically you're right.  The value is not cached in the card anymore.

However, the card is never shared with any other object. Each time `getCard` is called, it's immediately followed by a call to another getter, it's in order to get a specific value. Except in `updateSearchItemQA` of course, which is the function we are discussing. So nothing benefits of the caching. And `updateSearchItemQA` uses `mQA` to cache the question/answer directly and avoid using the cache of the card. It's `mQA` which is checked to know whether QA should be computed or not.

The only exception would be the copy constructor, which takes a CardCache to create another one. However, this constructor also copies `mQA`.

Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:2260]: Now that I write the explanation down, I admit it's far less obvious than how it sounded in my head while I made the change

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:2260]: I read these comments and think "On the one hand, technically you're right. The value is not cached in the card anymore." - if we lose cache coherency I either want the cache removed in the same PR or I want it invalidated somehow so it's clear it needs repopulation, or I want it maintained. Usually best to maintain it unless for performance reasons it's best to delay the hit and invalidate. Either way seems like it should be addressed directly, not obliquely with a long line of reasoning about why it's okay the cache is incoherent

**Comment by Arthur-Milchior** [comment]
@mikehardy I don't know what more feedback I can give you right now. I'm not stating that's a perfect answer, but I'll need help to understand your remark

File: AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/CardBrowser.java:2645]: > The only difference in this part of the code is that I don't put the value back in a map that is going to be deleted as soon as the method returns.

Can you guarantee for the future that the map will be deleted as soon as the method returns? Cache coherency errors are some of the most difficult to fix, and I'm very uncomfortable when cached values are out of date for any reason.

I will be much more comfortable if you either purge the map completely ("invalidate the cache") so future consumers are protected by some sort of fail fast behavior or you put the value back in the map ("make the cache coherent").

**Comment by Arthur-Milchior** [comment]
I've just looked at upstream code and seen that there is no hashmap anymore, instead there is a structure returned by rust code. So this change, not using a hashmap but a typed structure, will occur anyway when rust is here.
This change being NF, I can wait for rust without any trouble, so I'll just close it
